### ISSUE TITLE
Add possibility to act on warnings

### ISSUE THREAD

**Issue by kronn** [brigade/overcommit#288]
While introducing overcommit into a project, I was asked if it is possible to pause (and possibly abort the commit) before opening the editor for the commit-message.

I realize that this might be problematic if one is not using the git CLI, but some IDE or wrapper-script that just depends on the exit-status of the hook.

I see the following approaches and would like to get some feedback before investing time into a pull-request:

1. make explicit pausing a config-option, possibly limited to instances were overcommit runs in a tty.
2. add the warnings to the .git/COMMIT_MESSAGE (or whatever the file is called that git displays in the $EDITOR)
3. add a dry-run option that would run overcommit on all files currently staged for commit.

What do you think about this?

**Comment by kronn** [comment]
For my own reference, I want to hook (SCNR) into this method: https://github.com/brigade/overcommit/blob/master/lib%2Fovercommit%2Fhook_runner.rb#L72

**Comment by sds** [comment]
Hey @kronn, thanks for starting a discussion.

Right now I'm having difficulty understanding what you're trying to accomplish, so forgive me if I ask you to repeat yourself a bit.

Your request&mdash;as I understand it&mdash;is asking whether Overcommit would be open to providing a way to abort a commit before the commit editor opens. What are examples of some conditions where this would be necessary?

Your list of potential approaches mentions something about adding "warnings" to the `.git/COMMIT_MESSAGE` file. This suggests that you want to abort opening the editor if a certain error/warning condition is met, which sounds like a pre-commit hook and is thus already possible with Overcommit (if any pre-commit hook fails, the commit message editor will not be opened).

Is your ultimate goal to be able to run pre-commit hooks against your staged files without actually initiating a commit? Why would you want that? (there may be a valid reason, I'm just not seeing it)

**Comment by kronn** [comment]
Oh. It seems I left out the backstory and therefore the motivation :-)

The intended workflow was a follows:

1. user works on older code 
2. user adds and commits via CLI
3. user shortly sees that there are warnings
4. the editor opens to write the commit message.
5. now the user has the nagging feeling something was wrong, nothing serious, but maybe…
6. user commit anyway, with a bad feeling
7. NOW the warnings are visible again in the shell-scrollback buffer
8. if the users sees a real problem that affects the committed parts, he amends the last commit
9. user resolves the other warnings manually in different commits.

In the concrete case, the user was working on a codebase that just had puppet-lint added to a bigger codebase. so, naturally, there were a lot of warnings in file he touched, but not in the changes he made. Since there was an adaption period, some of the warnings were in the code he changed.

His concrete question to me was: Can I instruct overcommit to pause if there are warnings before opening the editor? So that I can possibly abort and change my code earlier?

As stated above, simply pausing might be difficult for interoperability, so I outlined the solutions that came to my mind to help with the above motivational backstory.

So, to answer your last question

> Is your ultimate goal to be able to run pre-commit hooks against your staged files without actually initiating a commit? 

Yes, this would be one possible solution.

> Why would you want that?

The reason would be to incrementally align the codebase to the rules outlined by puppet-lint (or rubocop or any other linter/stylechecker).

The problem is: I don't want to abort always, but sometimes there are "linting" issues which are in fact bugs (like inserting a variable in a non-interpolated string). So just raising every warning to failure might be a solution for one person, but is hurting the acceptance of overcommit and puppet-lint in the team.

In case you're wondering: yes, I'm looking for a technical solution to a problem that is part laziness and part social problem. But that describes software in general, right?

**Comment by sds** [comment]
Thanks for taking the time to describe your situation, @kronn. I definitely understand different teams have different processes and desires for how they want to use tools, and it's always difficult to build a general solution that satisfies all workflows.

My gut-feeling answer to this is that you are not using the tool as intended. Warnings are just that: warnings. They are not show stoppers, and you can always fix them at a later time in another commit. The use case you describe&mdash;as I understand it&mdash;seems to want to make warnings act like errors.

I want to address each of your original three suggestions below so you get an idea of where I'm coming from. I think some are definitely reasonable suggestions, but the implementation details have me wondering if the juice is worth the squeeze.

---

> 1. make explicit pausing a config-option, possibly limited to instances were overcommit runs in a tty.

I don't see what adding a pause in the presence of warnings gets you here other than an additional annoyance. You now need to wait for the pause to finish before moving on to writing your commit message. Your one developer will appreciate it, but the rest will be annoyed that errors they don't care about force them to wait. I can understand that there may be a strong desire to not introduce new warnings, but if you _really_ cared about not introducing them, you would configure them to report as errors. In my mind, it really is that simple.

Also, using the presence of a TTY to determine whether to insert the pause won't work, since all git hooks don't have access to a TTY in the first place (the STDIN stream is either closed or used to pipe in information passed by `git` itself to the hook, depending on the hook type, e.g. `post-rewrite`).

> 2. add the warnings to the .git/COMMIT_MESSAGE

On the surface I'm open to this suggestion as it solves your problem as I see it. However, from an implementation perspective I'm not happy with the road it leads us down. Ideally it would be implemented as a `prepare-commit-msg` hook, but then it would somehow need to be able to get the results of the previous hook run to insert into the commit message, which would require those results be saved to a file (since `git` itself invokes the hooks in separate processes). Communication between hook types via file strikes me as odd and inelegant (but it could be done).

> 3. add a dry-run option that would run overcommit on all files currently staged for commit.

This _almost_ already exists with the `--run` flag, but it executes on your entire repository, and would therefore be slower than running just against the changed files. 

You could create a new `HookContext` (similar to [`RunAll`](https://github.com/brigade/overcommit/blob/v0.28.0/lib/overcommit/hook_context/run_all.rb)) that works like the `PreCommit` context but is invoked via command-line flag instead of as an actual hook, so that the hooks all run but the editor is never opened. It's doable, but then you have to remember to use this special `--dry-run` flag in the first place, which seems silly since you end up running the same checks again when you `git commit` later.

---

To summarize, I think your suggestion of adding the warnings to the prepared commit message best solves your problem, but I'm not a fan of having hooks create files so they can relay results to each other. However, if a pull request adding this feature doesn't turn out to be hideous and is well-tested, I would be willing to merge it.

My preferred solution: switch the warnings to errors and fix them. :)

**Comment by kronn** [comment]
I have to think if the "solution" to collect warnings in a file and add them to the message is really worth the effort. From what I understand now, the `.git/COMMIT_MESSAGE` is only created once all previous hook signal "okay, proceed". Rather naïvely, i was hoping to just append to the already existing file. From what I learned now, this is not possible (fluxcapacitors aside).

I will discuss this with some colleagues to see what they as users of git and overcommit think about this. Personally, I tend to either accepts warnings as mere warnings now or treat all warnings as failures and fix them.

Let's see if - as you put it - the juice is worth the squeeze.

**Comment by kronn** [comment]
I discussed this and *drumroll* will solve this social problem not with technology. Thank you for the in-depth discussion about this.
